### PR TITLE
Add Junit report support

### PR DESCRIPTION
**Pull request by andrewegel**
Long standing request internally to spit out an XML report in junit format so it can be slurped up by CI-like systems (Jenkins).

### Testing

I modified the `check_runner_report_xunit` test to spit out an XML under `/tmp`, and ran it through `xmllint` against the Junit.xsd (https://llg.cubic.org/docs/junit/) schema for validation:

```
aegelhofer@Andrew-Egelhofer's- ducktape % xmllint --schema /tmp/ducktape-resuls/xunit.xsd /tmp/ducktape-resuls/report.xml --format
<?xml version="1.0"?>
<testsuites disabled="0" errors="0" failures="1" name="ducktape" tests="4" time="0.87104511261">
  <testsuite disabled="0" errors="0" failures="1" name="tests.runner.resources.test_thingy" skipped="2" tests="4">
    <testcase assertions="" classname="TestThingy" name="tests.runner.resources.test_thingy.TestThingy.test_pi" status="pass" time="0.00725317001343"/>
    <testcase assertions="" classname="TestThingy" name="tests.runner.resources.test_thingy.TestThingy.test_ignore1" status="ignore" time="0.0">
      <skipped/>
    </testcase>
    <testcase assertions="" classname="TestThingy" name="tests.runner.resources.test_thingy.TestThingy.test_ignore2.x=5" status="ignore" time="0.0">
      <skipped/>
    </testcase>
    <testcase assertions="" classname="TestThingy" name="tests.runner.resources.test_thingy.TestThingy.test_failure" status="fail" time="0.00793695449829">
      <failure message="This failed">This failed
Traceback (most recent call last):
  File "/Users/aegelhofer/workspace/ducktape/ducktape/tests/runner_client.py", line 134, in run
    data = self.run_test()
  File "/Users/aegelhofer/workspace/ducktape/ducktape/tests/runner_client.py", line 192, in run_test
    return self.test_context.function(self.test)
  File "/Users/aegelhofer/workspace/ducktape/tests/runner/resources/test_thingy.py", line 40, in test_failure
    raise Exception("This failed")
Exception: This failed
</failure>
    </testcase>
  </testsuite>
</testsuites>
/tmp/ducktape-resuls/report.xml validates
```

### CODE REVIEW

**Comment by rkgupta75** [comment]
I see we are using terminology "skipped" and "Ignore" interchangeably. Not sure if it is historic; rest LGTM.

File: tests/runner/check_runner.py [hunk 1]
```diff
@@ -68,20 +70,33 @@ def check_simple_run(self):
...
+        mock_cluster = LocalhostCluster(num_nodes=1000)
+        session_context = tests.ducktape_mock.session_context()
+        test_methods = [TestThingy.test_pi, TestThingy.test_ignore1, TestThingy.test_ignore2, TestThingy.test_failure]
+        ctx_list = self._do_expand(test_file=TEST_THINGY_FILE, test_class=TestThingy, test_methods=test_methods,
+                                   cluster=mock_cluster, session_context=session_context)
+        runner = TestRunner(mock_cluster, session_context, Mock(), ctx_list)
+
+        results = runner.run_all_tests()
+        XUnitReporter(results).report()
+        assert os.path.exists(os.path.join(session_context.results_dir, "report.xml"))
```
Comment (stan-confluent) [inline tests/runner/check_runner.py:98]: Would be good to also test that the result actually matches what you expect it to generate, not just that the file exists.

Comment (andrewegel) [reply tests/runner/check_runner.py:98]: Sure I'll add in some xpath asserts here...

Comment (stan-confluent) [reply tests/runner/check_runner.py:98]: Thanks! You could do xpath; or load this doc back into python objects and make asserts that way; or prepare a "golden" output doc, load both the doc generated in the test and the golden one in python and just `==` the resulting dicts.

Comment (andrewegel) [reply tests/runner/check_runner.py]: I opted for a more unit-driven approach with xpath (ie: unit test the number of testcase types in the XML doc).

**Comment by andrewegel** [comment]
> I see we are using terminology "skipped" and "Ignore" interchangeably. Not sure if it is historic; rest LGTM.

In ducktape terms its `ignored`, in Junit its `skipped` So thats the mapping you're seeing here - Junit also models tests as failures (`2/2 == 1`) or errors (`2/0` raises a divide by zero exception)  - ducktape treats both as "failures"

File: ducktape/tests/reporter.py [hunk 2]
```diff
@@ -17,6 +17,7 @@
 import json
 import os
 import shutil
+import xml.etree.ElementTree as ET
```
Comment (stan-confluent) [inline ducktape/tests/reporter.py:20]: Is xml already a dependency in setup.py?

File: ducktape/tests/reporter.py [hunk 3]
```diff
@@ -149,6 +150,65 @@ def report(self):
...
+            failures=str(self.results.num_failed)
+        ))
+        for module_name, testsuite in testsuites.items():
+            xml_testsuite = ET.SubElement(root, 'testsuite', attrib=dict(
+                name=module_name, tests=str(testsuite['tests']), disabled="0",
+                errors="0", failures=str(testsuite['failures']), skipped=str(testsuite['skipped'])
+            ))
+            for test in testsuite['testcases']:
+                xml_testcase = ET.SubElement(xml_testsuite, 'testcase', attrib=dict(
+                    name=test.test_id, classname=test.cls_name, time=str(test.run_time_seconds),
```
[the remark this answers is not in the document]
Comment (andrewegel) [reply ducktape/tests/reporter.py:194]: Fair ask.

[the remark this answers is not in the document]
Comment (andrewegel) [reply ducktape/tests/reporter.py]: Updated to just (essentially) be the method tester. ie:

`tests.runner.resources.test_thingy.TestThingy.test_ignore1` becomes `test_ignore1`
